wrangler-transform/src/main/java/io/cdap/wrangler/WranglerErrorUtil.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright © 2025 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.wrangler;
+
+import com.google.common.base.Throwables;
+import io.cdap.cdap.api.exception.ErrorCategory;
+import io.cdap.cdap.api.exception.ErrorType;
+import io.cdap.cdap.api.exception.ErrorUtils;
+import io.cdap.cdap.api.exception.ProgramFailureException;
+import io.cdap.wrangler.api.DirectiveLoadException;
+import io.cdap.wrangler.api.DirectiveParseException;
+import io.cdap.wrangler.api.RecipeException;
+import java.util.List;
+
+/**
+ * Error util file to handle exceptions caught in Wrangler plugin
+ */
+public final class WranglerErrorUtil {
+
+  private WranglerErrorUtil() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  public static ProgramFailureException getProgramFailureExceptionDetailsFromChain(Throwable e,
+      String errorMessage, ErrorType errorType) {
+    if (e != null) {
+      List<Throwable> causalChain = Throwables.getCausalChain(e);
+      for (Throwable t : causalChain) {
+        if (t instanceof ProgramFailureException) {
+          // Avoid double wrap
+          return (ProgramFailureException) t;
+        }
+        if (t instanceof DirectiveLoadException || t instanceof DirectiveParseException
+            || t instanceof RecipeException || t instanceof PreconditionException) {
+          return getProgramFailureException((Exception) t,
+              errorType != null ? errorType : ErrorType.USER, false);
```

Comment:
based on exception, there can be different `subCategory` like `Parsing-Directive`, `Loading-Directive`, `Executing-Recipe` & `Precondition`